### PR TITLE
Ensure that we are actually asking for optimistic locking for moves

### PR DESCRIPTION
**Pull request by daines**
- Make sure the lastModifiedDate parameter is actually added when attempting to modify a datastream
- Retrieve a datastream's metadata before retrieving its content so that we have its last modified date to millisecond resolution, rather than second resolution as available from the Last-Modified header
- Clean up debug logging in move methods (removeChildren, addChildren, cleanupRemovedChildren)

### CODE REVIEW

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java [hunk 1]
```diff
@@ -623,29 +625,42 @@ public String getLastModified() {
...
+			
+			while (true) {
+		
+				edu.unc.lib.dl.fedora.types.Datastream datastream = managementClient.getDatastream(pid, datastreamName);
+			
+				if (datastream == null) {
+					return null;
 				}
+				
+				log.debug("Got datastream, attempting to get dissemination version with create date " + managementClient.getDatastream(pid, datastreamName).getCreateDate());
```
Comment (bbpennel) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: Guessing this is just in here for testing?  Probably don't want it re-retrieval the datastream profile.

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java [hunk 2]
```diff
@@ -623,29 +625,42 @@ public String getLastModified() {
...
-					lastModified = header.getValue();
-					break;
+		
+		log.debug("Attempting to get datastream " + datastreamName + " for object " + pid);
+		
+		try {
+			
+			while (true) {
+		
+				edu.unc.lib.dl.fedora.types.Datastream datastream = managementClient.getDatastream(pid, datastreamName);
```
Comment (bbpennel) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: This is likely the better way to check for the presence of a datastream.  Do you have an idea of if it has a comparable execution time to the triplestore check in the previous version?

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java [hunk 3]
```diff
@@ -623,29 +625,42 @@ public String getLastModified() {
...
+			
+				if (datastream == null) {
```
Comment (bbpennel) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: Are both this conditional and the NotFoundException catch necessary?  Is the exception not being thrown every time?

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java [hunk 4]
```diff
@@ -623,29 +625,42 @@ public String getLastModified() {
...
+				
+				log.debug("Got datastream, attempting to get dissemination version with create date " + managementClient.getDatastream(pid, datastreamName).getCreateDate());
+			
+				try {
+			
+					MIMETypedStream mts = accessClient.getDatastreamDissemination(pid, datastreamName, datastream.getCreateDate());
+				
+					try (ByteArrayInputStream bais = new ByteArrayInputStream(mts.getStream())) {
+						Document dsDoc = new SAXBuilder().build(bais);
+						return new DatastreamDocument(dsDoc, datastream.getCreateDate());
```
Comment (bbpennel) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: So it ended up being okay to use the timestamp returned by Fedora in the getDatastream request?  It has enough precision to be useful for optimistic locking, unlike the modified header in the getDatastreamDissemination call?

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java (hunk 2 above)
Comment (daines) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: I'm not sure about execution time. This isn't only a check for existence, though. We need the "create date" for the latest version of the datastream so we can set lastModified on the DatastreamDocument object we return.

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java (hunk 3 above)
Comment (daines) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: It seems to always be null, but getDatastream does throw FedoraException.

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java (hunk 1 above)
Comment (daines) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: oops, should just be datastream.getCreateDate()

File: persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java (hunk 4 above)
Comment (daines) [inline persistence/src/main/java/edu/unc/lib/dl/services/DigitalObjectManagerImpl.java]: Yes.

**Pull request merged**